lib/miq_expression.rb

```diff
@@ -1621,6 +1621,25 @@ def custom_attribute_columns_and_models(expression = nil)
     end
   end
 
+  def fields(expression = exp, ret = [])
```

Comment:
I'd recommend *not* passing state (ret param in this case) through the recursion.  If written functionally, then you don't need it.  That is, if the method returns the fields of the given expression, then the caller can do the appending...

```ruby
def fields(expression = exp)
  case expression
  when Array
    expression.flat_map { |x| fields(x) }
  when Hash
    return [] if expression.empty?

    if (val = expression["field"])
      ret = []

      ret << Field.parse(val) if Field.valid_field?(val)
      val = expression["value"]
      ret << Field.parse(val) if Field.valid_field?(val)

      ret
    else
      fields(expression.values)
    end
  end
end
```